File: wheel/src/main/java/dev/yong/wheel/cache/CacheManager.java
package dev.yong.wheel.cache;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.io.RandomAccessFile;
import java.io.Serializable;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicLong;

/**
 * @author coderyong
 */
class CacheManager {
    private final AtomicLong mCacheSize;
    private final AtomicInteger mCacheCount;
    private File mCacheDir;
    private final long mMaxSize;
    private final int mMaxCount;
    private final Map<String, File> mCacheFiles = Collections
            .synchronizedMap(new HashMap<>());
    private final Map<File, Long> mLastModified = Collections
            .synchronizedMap(new HashMap<>());

    CacheManager(File cacheDir, long maxSize, int maxCount) {
        this.mCacheDir = cacheDir;
        this.mMaxSize = maxSize;
        this.mMaxCount = maxCount;
        mCacheSize = new AtomicLong();
        mCacheCount = new AtomicInteger();
        calculateCacheSizeAndCacheCount();
    }

    /**
     * 计算 cacheSize和cacheCount
     */
    private void calculateCacheSizeAndCacheCount() {
        Executors.newSingleThreadExecutor().execute(() -> {
            int size = 0;
            int count = 0;
            File[] cachedFiles = mCacheDir.listFiles();
            if (cachedFiles != null) {
                for (File cachedFile : cachedFiles) {
                    size += cachedFile.length();
                    count += 1;
                    mLastModified.put(cachedFile, cachedFile.lastModified());
                    mCacheFiles.put(getFileKey(cachedFile), cachedFile);
                }
                mCacheSize.set(size);
                mCacheCount.set(count);
            }
        });
    }

    private File get(String key) {
        if (!isObsolete(key)) {
            File file = getFile(key);
            file.setLastModified(System.currentTimeMillis());
            mLastModified.put(file, file.lastModified());
            return file;
        }
        return null;
    }

    private File newFile(String key, long timeout, TimeUnit unit) {
        File file = getFile(key);
        if (file != null) {
            file.deleteOnExit();
        }
        String fileName = key.hashCode() + "";
        if (timeout > -1) {
            long time = System.currentTimeMillis() + unit.toMillis(timeout);
            fileName = key.hashCode() + "_" + time;
        }
        return new File(mCacheDir, fileName);
    }

    private String getFileKey(File cachedFile) {
        return cachedFile.getName().split("_")[0];
    }

    private File getFile(String key) {
        return mCacheFiles.get(key.hashCode() + "");
    }

    private long getFileTimeout(File cachedFile) {
        String name = cachedFile.getName();
        String timeStr = "-1";
        if (name.contains("_")) {
            timeStr = cachedFile.getName().split("_")[1];
        }
        try {
            return Long.parseLong(timeStr);
        } catch (Exception e) {
            return -1;
        }
    }

    /**
     * 添加缓存文件
     *
     * @param key     缓存文件key
     * @param value   缓存数据
     * @param timeout 设置超时删除
     * @param unit    时间单位{@link TimeUnit}
     */
    public void put(String key, Serializable value, long timeout, TimeUnit unit) {
        ByteArrayOutputStream out;
        ObjectOutputStream oos = null;
        try {
            out = new ByteArrayOutputStream();
            oos = new ObjectOutputStream(out);
            oos.writeObject(value);
            byte[] data = out.toByteArray();
            put(key, data, timeout, unit);
            out.close();
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            try {
                if (oos != null) {
                    oos.close();
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

    /**
     * 添加缓存文件
     *
     * @param key     缓存文件key
     * @param value   缓存数据
     * @param timeout 设置超时删除
     * @param unit    时间单位{@link TimeUnit}
     */
    public void put(String key, byte[] value, long timeout, TimeUnit unit) {
        File file = newFile(key, timeout, unit);
        writeDataToFile(file, value);
        int curCacheCount = mCacheCount.get();
        while (curCacheCount + 1 > mMaxCount) {
            long freedSize = removeNext();
            mCacheSize.addAndGet(-freedSize);

            curCacheCount = mCacheCount.addAndGet(-1);
        }
        mCacheCount.addAndGet(1);

        long valueSize = file.length();
        long curCacheSize = mCacheSize.get();
        while (curCacheSize + valueSize > mMaxSize) {
            long freedSize = removeNext();
            curCacheSize = mCacheSize.addAndGet(-freedSize);
        }
        mCacheSize.addAndGet(valueSize);

        file.setLastModified(System.currentTimeMillis());
        mLastModified.put(file, file.lastModified());
        mCacheFiles.put(getFileKey(file), file);
    }

    /**
     * 将数据写入到文件
     *
     * @param file  目标文件
     * @param value 缓存数据
     */
    private void writeDataToFile(File file, byte[] value) {
        FileOutputStream out = null;
        try {
            out = new FileOutputStream(file);
            out.write(value);
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            if (out != null) {
                try {
                    out.flush();
                    out.close();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }
    }

    /**
     * 获取String数据
     *
     * @param key 缓存文件key
     * @return String 数据
     */
    public String getString(String key) {
        return (String) getSerializable(key);
    }

    /**
     * 获取序列化数据
     *
     * @param key 缓存文件key
     * @return Serializable 数据
     */
    Serializable getSerializable(String key) {
        byte[] data = getByte(key);
        if (data != null) {
            ByteArrayInputStream in = null;
            ObjectInputStream ois = null;
            try {
                in = new ByteArrayInputStream(data);
                ois = new ObjectInputStream(in);
                return (Serializable) ois.readObject();
            } catch (Exception e) {
                e.printStackTrace();
            } finally {
                try {
                    if (in != null) {
                        in.close();
                    }
                    if (ois != null) {
                        ois.close();
                    }
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }
        return null;
    }

    /**
     * 获取 byte 数据
     *
     * @param key 缓存文件key
     * @return byte 数据
     */
    byte[] getByte(String key) {
        RandomAccessFile accessFile = null;
        try {
            File file = get(key);
            if (file == null || !file.exists()) {
                return null;
            }
            accessFile = new RandomAccessFile(file, "r");
            byte[] byteArray = new byte[(int) accessFile.length()];
            accessFile.read(byteArray);
            return byteArray;
        } catch (Exception e) {
            e.printStackTrace();
            return null;
        } finally {
            if (accessFile != null) {
                try {
                    accessFile.close();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }
    }

    /**
     * 判断缓存的String数据是否到期
     *
     * @param key 文件保存的Key
     * @return true：到期了 false：还没有到期
     */
    boolean isObsolete(String key) {
        File file = getFile(key);
        if (file == null) {
            return true;
        }
        long timeout = getFileTimeout(file);
        boolean isObsolete = timeout != -1 && timeout < System.currentTimeMillis();
        if (isObsolete) {
            if (file.delete()) {
                mCacheFiles.remove(key);
                mLastModified.remove(file);
            } else {
                isObsolete = false;
            }
        }
        return isObsolete;
    }

    /**
     * 移除指定缓存数据
     *
     * @param key 缓存文件key
     * @return 是否移除成功
     */
    boolean remove(String key) {
        return getFile(key).delete();
    }

    /**
     * 清除缓存文件夹
     */
    boolean clear() {
        mLastModified.clear();
        mCacheFiles.clear();
        mCacheSize.set(0);
        return deleteFile(mCacheDir);
    }

    /**
     * 遍历删除文件
     *
     * @param file 文件目录
     */
    private boolean deleteFile(File file) {
        if (file == null) {
            return false;
        }
        if (file.isFile()) {
            return file.delete();
        } else {
            File[] files = file.listFiles();
            if (files != null) {
                for (File f : files) {
                    if (!deleteFile(f)) {
                        return false;
                    }
                }
            }
            return true;
        }
    }

    /**
     * 移除旧的文件
     *
     * @return 移除文件的大小
     */
    private long removeNext() {
        if (mLastModified.isEmpty()) {
            return 0;
        }

        File oldestFile = null;
        Long oldestModified = null;
        Set<Map.Entry<File, Long>> entries = mLastModified.entrySet();
        synchronized (mLastModified) {
            for (Map.Entry<File, Long> entry : entries) {
                if (oldestFile == null) {
                    oldestFile = entry.getKey();
                    oldestModified = entry.getValue();
                } else {
                    Long lastModified = entry.getValue();
                    if (lastModified < oldestModified) {
                        oldestFile = entry.getKey();
                        oldestModified = lastModified;
                    }
                }
            }
        }
        if (oldestFile != null) {
            if (oldestFile.delete()) {
                mLastModified.remove(oldestFile);
                mCacheFiles.remove(getFileKey(oldestFile));
            }
        }
        return 0;
    }
}

File: wheel/src/main/java/dev/yong/wheel/cache/CacheManager.kt
package dev.yong.wheel.cache

import java.io.*
import java.util.*
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.atomic.AtomicLong

/**
 * @author coderyong
 */
internal class CacheManager(
    private val mCacheDir: File,
    private val mMaxSize: Long,
    private val mMaxCount: Int
) {
    private val mCacheSize: AtomicLong = AtomicLong()
    private val mCacheCount: AtomicInteger = AtomicInteger()
    private val mCacheFiles = Collections
        .synchronizedMap(HashMap<String, File>())
    private val mLastModified = Collections
        .synchronizedMap(HashMap<File, Long>())

    /**
     * 计算 cacheSize和cacheCount
     */
    private fun calculateCacheSizeAndCacheCount() {
        Executors.newSingleThreadExecutor().execute {
            var size = 0
            var count = 0
            val cachedFiles = mCacheDir.listFiles()
            if (cachedFiles != null) {
                for (cachedFile in cachedFiles) {
                    size += cachedFile.length().toInt()
                    count += 1
                    mLastModified[cachedFile] = cachedFile.lastModified()
                    mCacheFiles[getFileKey(cachedFile)] = cachedFile
                }
                mCacheSize.set(size.toLong())
                mCacheCount.set(count)
            }
        }
    }

    private operator fun get(key: String): File? {
        if (!isObsolete(key)) {
            val file = getFile(key)
            mLastModified[file!!] =
                if (file.setLastModified(System.currentTimeMillis())) file.lastModified() else System.currentTimeMillis()
            return file
        }
        return null
    }

    private fun newFile(key: String, timeout: Long, unit: TimeUnit): File {
        val file = getFile(key)
        file?.deleteOnExit()
        var fileName = key.hashCode().toString() + ""
        if (timeout > -1) {
            val time = System.currentTimeMillis() + unit.toMillis(timeout)
            fileName = key.hashCode().toString() + "_" + time
        }
        return File(mCacheDir, fileName)
    }

    private fun getFileKey(cachedFile: File): String {
        return cachedFile.name.split("_".toRegex()).toTypedArray()[0]
    }

    private fun getFile(key: String): File? {
        return mCacheFiles[key.hashCode().toString() + ""]
    }

    private fun getFileTimeout(cachedFile: File): Long {
        val name = cachedFile.name
        var timeStr = "-1"
        if (name.contains("_")) {
            timeStr = cachedFile.name.split("_".toRegex()).toTypedArray()[1]
        }
        return try {
            timeStr.toLong()
        } catch (e: Exception) {
            -1
        }
    }

    /**
     * 添加缓存文件
     *
     * @param key     缓存文件key
     * @param value   缓存数据
     * @param timeout 设置超时删除
     * @param unit    时间单位[TimeUnit]
     */
    fun put(key: String, value: Serializable, timeout: Long, unit: TimeUnit) {
        val out: ByteArrayOutputStream
        var oos: ObjectOutputStream? = null
        try {
            out = ByteArrayOutputStream()
            oos = ObjectOutputStream(out)
            oos.writeObject(value)
            val data = out.toByteArray()
            put(key, data, timeout, unit)
            out.close()
        } catch (e: Exception) {
            e.printStackTrace()
        } finally {
            try {
                oos?.close()
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
    }

    /**
     * 添加缓存文件
     *
     * @param key     缓存文件key
     * @param value   缓存数据
     * @param timeout 设置超时删除
     * @param unit    时间单位[TimeUnit]
     */
    fun put(key: String, value: String, timeout: Long, unit: TimeUnit) {
        put(key, value.toByteArray(), timeout, unit)
    }


    /**
     * 添加缓存文件
     *
     * @param key     缓存文件key
     * @param value   缓存数据
     * @param timeout 设置超时删除
     * @param unit    时间单位[TimeUnit]
     */
    fun put(key: String, value: ByteArray, timeout: Long, unit: TimeUnit) {
        val file = newFile(key, timeout, unit)
        writeDataToFile(file, value)
        var curCacheCount = mCacheCount.get()
        while (curCacheCount + 1 > mMaxCount) {
            val freedSize = removeNext()
            mCacheSize.addAndGet(-freedSize)
            curCacheCount = mCacheCount.addAndGet(-1)
        }
        mCacheCount.addAndGet(1)
        val valueSize = file.length()
        var curCacheSize = mCacheSize.get()
        while (curCacheSize + valueSize > mMaxSize) {
            val freedSize = removeNext()
            curCacheSize = mCacheSize.addAndGet(-freedSize)
        }
        mCacheSize.addAndGet(valueSize)
        mLastModified[file] =
            if (file.setLastModified(System.currentTimeMillis())) file.lastModified() else System.currentTimeMillis()
        mCacheFiles[getFileKey(file)] = file
    }

    /**
     * 将数据写入到文件
     *
     * @param file  目标文件
     * @param value 缓存数据
     */
    private fun writeDataToFile(file: File, value: ByteArray) {
        var out: FileOutputStream? = null
        try {
            out = FileOutputStream(file)
            out.write(value)
        } catch (e: Exception) {
            e.printStackTrace()
        } finally {
            if (out != null) {
                try {
                    out.flush()
                    out.close()
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            }
        }
    }

    /**
     * 获取String数据
     *
     * @param key 缓存文件key
     * @return String 数据
     */
    fun getString(key: String): String? {
        val bytes = getByte(key)
        return if (bytes == null) null else String(bytes)
    }

    /**
     * 获取序列化数据
     *
     * @param key 缓存文件key
     * @return Serializable 数据
     */
    fun getSerializable(key: String): Serializable? {
        val data = getByte(key)
        if (data != null) {
            var `in`: ByteArrayInputStream? = null
            var ois: ObjectInputStream? = null
            try {
                `in` = ByteArrayInputStream(data)
                ois = ObjectInputStream(`in`)
                return ois.readObject() as Serializable
            } catch (e: Exception) {
                e.printStackTrace()
            } finally {
                try {
                    `in`?.close()
                    ois?.close()
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            }
        }
        return null
    }

    /**
     * 获取 byte 数据
     *
     * @param key 缓存文件key
     * @return byte 数据
     */
    fun getByte(key: String): ByteArray? {
        var accessFile: RandomAccessFile? = null
        return try {
            val file = get(key)
            if (file == null || !file.exists()) {
                return null
            }
            accessFile = RandomAccessFile(file, "r")
            val byteArray = ByteArray(accessFile.length().toInt())
            accessFile.read(byteArray)
            byteArray
        } catch (e: Exception) {
            e.printStackTrace()
            null
        } finally {
            if (accessFile != null) {
                try {
                    accessFile.close()
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            }
        }
    }

    /**
     * 判断缓存的String数据是否到期
     *
     * @param key 文件保存的Key
     * @return true：到期了 false：还没有到期
     */
    fun isObsolete(key: String): Boolean {
        val file = getFile(key)
        if (file == null || !file.exists()) {
            return true
        }
        val timeout = getFileTimeout(file)
        var isObsolete = timeout != -1L && timeout < System.currentTimeMillis()
        if (isObsolete) {
            if (file.delete()) {
                mCacheFiles.remove(key)
                mLastModified.remove(file)
            } else {
                isObsolete = false
            }
        }
        return isObsolete
    }

    /**
     * 移除指定缓存数据
     *
     * @param key 缓存文件key
     * @return 是否移除成功
     */
    fun remove(key: String): Boolean {
        val file = getFile(key)
        return file == null || !file.exists() || file.delete()
    }

    /**
     * 清除缓存文件夹
     */
    fun clear(): Boolean {
        mLastModified.clear()
        mCacheFiles.clear()
        mCacheSize.set(0)
        return deleteFile(mCacheDir)
    }

    /**
     * 遍历删除文件
     *
     * @param file 文件目录
     */
    private fun deleteFile(file: File?): Boolean {
        if (file == null) {
            return false
        }
        return if (file.isFile) {
            file.delete()
        } else {
            val files = file.listFiles()
            if (files != null) {
                for (f in files) {
                    if (!deleteFile(f)) {
                        return false
                    }
                }
            }
            true
        }
    }

    /**
     * 移除旧的文件
     *
     * @return 移除文件的大小
     */
    private fun removeNext(): Long {
        if (mLastModified.isEmpty()) {
            return 0
        }
        var oldestFile: File? = null
        var oldestModified: Long? = null
        val entries: Set<Map.Entry<File, Long>> = mLastModified.entries
        synchronized(mLastModified) {
            for ((key, lastModified) in entries) {
                if (oldestFile == null) {
                    oldestFile = key
                    oldestModified = lastModified
                } else {
                    if (lastModified < oldestModified!!) {
                        oldestFile = key
                        oldestModified = lastModified
                    }
                }
            }
        }
        if (oldestFile != null) {
            if (oldestFile!!.delete()) {
                mLastModified.remove(oldestFile)
                mCacheFiles.remove(getFileKey(oldestFile!!))
            }
        }
        return 0
    }

    init {
        calculateCacheSizeAndCacheCount()
    }
}